README.md
# gitbook-plugin-iuap-desigin

## 说明

gitbook-plugin-iuap-desigin是基于gitbook的扩展功能插件。

## 功能介绍

- 支持左侧目录展开
- 引入 `iuap-desigin` 相关资源时，可同时将生成的 `html` 完美融入 `iuap-desigin` 官网。

## 使用

```
npm install gitbook-plugin-iuap-desigin --save-dev
```

`gitbook` 的 `book.json` 配置示例

```
"language":"zh-hans",
"plugins":[
  "iuap-desigin"
],
```


CHANGELOG.md
# 更新日志

2016-06-29 v1.2.9

- 解决menu菜单抖动、滚动条无法浮现bug

2016-06-30 v1.2.10

- 美化menu菜单滚动条效果

2016-06-30 v1.2.11

- 修改bug,去除冗余代码

2016-06-30 v1.2.12

- 增加页面刷新:progress,代码高亮现可正常显示
- 调整侧menu默认样式

2016-06-30 v1.2.13

- 修改左侧menu style偏移bug

2016-06-30 v1.2.14

- 左侧menu设定高度，消除fix滚动条bug

2016-06-30 v1.2.15

- 修改冲突

2016-06-30 v1.2.17

- 调整rat.js加载位置

2016-07-02 v1.2.18

- 修改锚点位置bug
- 修改侧导航划出目录错位bug
- 完善左侧菜单目录链接正则验证,确保页面正确跳转

2016-07-05 v1.2.19

- 修改模板，确保整站共用header/footer

2016-07-06 v1.2.20

- 完善跳转链接正则匹配，避免锚点链接后点击目录无法跳转bug

2016-07-06 v1.2.21

- 取消导航响应式样式控制,公共部分样式调整统一在style.css

2016-07-06 v1.2.22

- 调整左侧内容高度，避免部分内容不能正常显示

2016-07-06 v1.2.23

- 模板增加u.css,避免tooltip显示bug
- 增加u-grid.js,u-trree.js,grid.css,tree.css

2016-07-06 v1.2.24

- 调整左侧目录，防止抖动

2016-07-06 v1.2.25

- clear代码片段

2016-07-06 v1.2.26

- 修改1.0.3 tree.css 为 3.0.6

2016-07-11 v1.2.27

- 所有1.0.3资源迁移到最新的3.0.6


2016-07-12 v1.2.28

- 修改layout加载的u.css版本，调整冗余u.css

2016-07-12 v1.2.29

- 调整加载均指向最新版路径/latest

2016-07-12 v1.3.0

- 修改包裹style,script标签的显示(display:none)

2016-07-12 v1.3.1

- css增加font样式

2016-07-14 v1.3.2

- 调整内容区样式
- 调整layout引用路径
- 修改公共头尾路径

2016-07-15 v1.3.3

* 调整左导航高度及相关显示bug

2016-07-16 v1.3.4

* 处理ie9 history.statereplace bug

2016-07-17 v1.3.5

* 给予右侧内容区域最低高度,解决分割线长度问题bug

2016-07-19 v1.3.6

* 解决firefox跳转问题

2016-07-21 v1.3.7

* 针对ie8增加polyfill

2016-07-21 v1.3.8

* 增加搜索样式

2016-07-26 v1.3.9

* 修改表格th样式居中

2016-07-27 v1.3.10

* 修改h4级样式问题

2016-07-27 v1.3.11

* 修改左侧导航样式

2016-07-27 v1.3.12

* 补充修改左侧导航

2016-07-27 v1.3.13

* 版本调试

2016-07-27 v1.4.1

* 左导航增加resize事件
* 左导航高度绑定,避免底部留白
* 左导航样式调整

2016-08-01 v1.4.2

* 修改侧导航active样式

2016-08-01 v1.4.3

* 修改plugin插件被覆盖的样式

2016-08-01 v1.4.4

* 修改clock-picker样式

index.js
var cheerio = require('cheerio');
var _ = require('underscore');
var fs = require('fs');
var path = require('path');

module.exports = {
    book: {
        assets: ".",
    },
    hooks: {
        "page": function (page) {
            return page;
        }
    },
    website: {
        assets: './book',
        js: [
            'iuap-design.js'
        ],
        css: [
            'iuap-design.css'
        ],
        html: {
          /**
           * head 内加载的资源
           * @param  {[type]} current [description]
           * @return {[type]}         [description]
           */
          "head:end": function(current) {

            var pathStr = '';
            var ctx = 'http://design.yyuap.com/static/';

            // Polyfill
            pathStr = '<!--[if lte IE 8]>' +
              '<script src="//cdn.bootcss.com/html5shiv/3.7.3/html5shiv.min.js"></script>\r\n' +
              '<script src="//cdn.bootcss.com/respond.js/1.4.2/respond.min.js"></script>\r\n' +
              '<script src="http://design.yyuap.com/static/uui/latest/js/u-polyfill.js"></script>\r\n' +
              '<![endif]-->';

            // Path Array
            var pathArray=[
              'uui/latest/css/font-awesome.css',
              // 'uui/latest/css/u.css',
              // 'uui/latest/u-extend.css',
              'uui/latest/css/tree.css',
              'uui/latest/css/grid.css',
              'highlight/styles/atelier-plateau-light.css',
              'scrollbar/jquery.mCustomScrollbar.css',
              'nanoscroller/nanoscroller.css',
              'jquery/jquery-1.9.1.min.js',
              'raty/rating.js',
              'knockout/knockout-3.2.0.debug.js',
              'uui/latest/js/u.js'
            ];

            for(var i = 0, len = pathArray.length; i < len; i++){
              if ( /\.css$/.test(pathArray[i]) ) {
                pathStr += '<link rel="stylesheet" href="'
                  + ctx + pathArray[i] + '">\r\n';
              } else if ( /\.js$/.test(pathArray[i]) ) {
                pathStr += '<script src="' + ctx + pathArray[i] + '"></script>\r\n';
              }
            }

            return pathStr;
          },

          /**
           * body 内容体
           * @param  {[type]} current [description]
           * @return {[type]}         [description]
           */
          "body:start": function(current) {
            var title = this.options.pluginsConfig['iuap-design']['title'] || "";
            var desc = this.options.pluginsConfig['iuap-design']['desc'] || "";

            var headPath = path.join('../../src/common/','header.html');
            var headCont = fs.readFileSync(headPath, {encoding: "utf-8"});

            var banner = [
              '<div class="banner">',
                ' <div class="container">',
                '   <div class="banner-content">',
                '       <h1>' + title + '</h1>',
                  '         <p class="info">' + desc + '</p>',
                '   </div>',
                ' </div>',
              '</div>'
            ]

            var bannerHTML = banner.join('\r\n');

            return headCont + bannerHTML;
          },
          /**
           * body 尾部内容，gitbook插件之后加载
           * @param  {[type]} current [description]
           * @return {[type]}         [description]
           */
          "body:end": function(current) {

            var jsStr = '';
            var ctx = 'http://design.yyuap.com/static/';

            var footPath = path.join('../../src/common/','footer.html');
            var footCont = fs.readFileSync(footPath, {encoding: "utf-8"});

            var LibArray = [
              'knockout/knockout-3.2.0.debug.js',
              'scrollbar/jquery.mCustomScrollbar.concat.min.js',
              'nanoscroller/jquery.nanoscroller.js',
              'highlight/highlight.min.js',
              'uui/latest/js/u-polyfill.js',
              'uui/latest/js/u-tree.js',
              'uui/latest/js/u-grid.js'
            ];

            for ( var i = 0, len = LibArray.length; i < len; i++ ) {
              jsStr += '<script src="'+ ctx + LibArray[i] + '"></script>\r\n';
            }

            return footCont + jsStr;
          },
          /**
           * body 尾部后添加
           * @param  {[type]} argument [description]
           * @return {[type]}          [description]
           */
          "body:add": function (argument) {
            
            var scriptStr = ''
            var codeStr = [
              '$(".nano").nanoScroller();',
              'hljs.initHighlightingOnLoad();'
            ]

            scriptStr += '<script>';            
            for( var i=0, codeLen = codeStr.length; i < codeLen; i++ ) {
              scriptStr += '\r\n' + codeStr[i];
            }
            scriptStr += '</script>';

            return scriptStr;
          }
        }
    }
};


book/iuap-desigin.js
require( [ 'gitbook', 'jQuery', 'lodash' ], function ( gitbook, $, _ ) {

	/**
	 * [bind description]
	 * @param  {[type]} 'start'  [description]
	 * @param  {[type]} function (             [description]
	 * @return {[type]}          [description]
	 */
	gitbook.events.bind('start', function () {
	});

	/**
	 * 处理页面逻辑
	 * @param  {[type]} 'page.change' [description]
	 * @param  {[type]} function      (             [description]
	 * @return {[type]}               [description]
	 */
	gitbook.events.bind('page.change', function () {

		var $body = $('body');
		// 文档部分
		var $book = $('.book');
		// 文档左侧目录
		var $summary = $('.book-summary');
		// 文档右侧主体
		var $bookBody = $('.book-body');
		var $container = $('<div class="container"></div>');
		var $containerDiv = $('<div class="container-div"></div>');

		$book.append($container);
		$container.append($containerDiv);
		$containerDiv.append($summary);
		$containerDiv.append($bookBody);

		// 左侧目录修改
		var $summaryUl = $('ul',$summary);
		var $dividerLi = $('.divider',$summaryUl);
		var $dividerLiNext = $('.divider + li',$summaryUl);

		// 去掉介绍
		$('.summary > li').first().remove();

		// 删除下面横线及之后的li
		$dividerLi.remove();
		$dividerLiNext.remove();

		var $summaryAB = $('a b',$summary);

		// 去掉目录的编号
		$summaryAB.remove();

		// 将超链接放到page-wrapper的最后
		var $pageWrapper = $('.page-wrapper');
		var $prevA = $('.navigation-prev',$bookBody);
		var $nextA = $('.navigation-next',$bookBody);

		$pageWrapper.append($prevA);
		$pageWrapper.append($nextA);
	    var $html = $('html');
	    $html.css('font-size','62.5%');

	    $('body').css('display','block');

	    function bodyScrollFun(){
	    	var st = document.body.scrollTop || document.documentElement.scrollTop;
	    	var t = 215 - st;
	    	var t1 = 240 - st;

				if ( t > 0 ) t = 0;

				$('.book-summary').css('top', -1 * t + 'px');

				if ( t1 < 80 ) t1 = 30

				$('#anchors-navbar').css('top', t1 + 'px');
	    }

	    function bodyScroll(){
	    	document.body.onscroll = bodyScrollFun;
	    }

	    document.body.scrollTop = 0;
	    setInterval(bodyScroll,100)
	    bodyScrollFun();

	    var oH = document.body.offsetHeight;
	    var h = parseInt(oH) - 80;

	    $book.css('min-height',parseInt(oH) - 60 + 'px');
	    $containerDiv.css('min-height',parseInt(oH) - 120 + 'px');

	    /**
	     *  highlight
	     */
	    hljs.initHighlightingOnLoad();

		// mobile side menu init
	    (function(){
	    	function mobileNav(){
	        	var bodyWidth = document.body.offsetWidth;
	        	var eleBook = document.querySelectorAll('.book')[0];
	        	if(bodyWidth<=600) {
	        		eleBook.classList ? eleBook.classList.remove('with-summary') : eleBook.className.replace(new RegExp('(^|\\b)' + 'with-summary'.split(' ').join('|') + '(\\b|$)', 'gi'), ' ');
	        	}
	   		}
	   		mobileNav();
	   		window.onresize = mobileNav;
	    })();

	    /**
	     * [description] Main menu slide
	     * @return {[type]} [description]
	     */
	    (function(){

			var openOffcanvas = $('[data-offcanvas="open"]');
			var closeOffcanvas = $('[data-offcanvas="close"]');
			var offcanvasNav = $('.offcanvas-nav');
			openOffcanvas.click(function(){
				openOffcanvas.addClass('nav-open');
				offcanvasNav.addClass('open');
				$('body').append('<div class="offcanvas-backdrop"></div>');
			});
			closeOffcanvas.click(function(){
				openOffcanvas.removeClass('nav-open');
				offcanvasNav.removeClass('open');
				$('.offcanvas-backdrop').remove();
			});
			$(document).on('click', '.offcanvas-backdrop', function(){
				openOffcanvas.removeClass('nav-open');
				offcanvasNav.removeClass('open');
				$('.offcanvas-backdrop').remove();
			});
	    })();

	    var uDesign = {
	    	init: function() {
	    		this.buildTag();

					console.log( $.mCustomScrollbar );
					
	    	},
	    	/**
	    	 * 创建<style>,<script>标签
	    	 * @return {[type]} [description]
	    	 */
	    	buildTag: function() {
	    		var jsCode = document.querySelectorAll('.jstag');
	    		var cssCode = document.querySelectorAll('.csstag');
	    		var jsTag = document.createElement('script');
	    		var cssTag = document.createElement('style');

	    		for(var i=0, jsLen = jsCode.length; i<jsLen; i++) {
	    			jsTag.innerHTML += jsCode[i].textContent;
	    		}
	    		document.body.appendChild(jsTag);

	    		for(var j=0, cssLen = cssCode.length; j<cssLen; j++ ) {
	    			cssTag.innerHTML = cssCode[j].textContent;
	    		}
	    		document.head.appendChild(cssTag);
	    	}
	    }

	    uDesign.init();


	});






});


book/iuap-design.js
require( [ 'gitbook' ], function ( gitbook ) {

	window.onload = function() {
		var uGitpage = {
			init: function(){
				this.summaryInit();
			},
			/**
			 * DOM元素
			 */
			DOM:{
				"$summary": $('.book-summary'),
				"$bookInner": $('.body-inner'),
				"footer":document.getElementsByTagName('footer')[0],
			},
			/**
			 * 左侧menu初始化
			 */
			summaryInit:function(){
				var DOM = this.DOM;

			    // 获取浏览器高度
				var browserH = document.documentElement.clientHeight;

	    		// 左侧目录高度设置
	    		var sumaryHeightFun = function() {
	    			// DOM.$bookInner.css('min-height',browserH + 'px');
	    			var bookInner = document.querySelectorAll('.body-inner')[0];
	    			bookInner.style.minHeight = browserH + 'px';
	    			var bookInnerHeight = parseInt(getComputedStyle(bookInner)['height']);

	    			// 判断footer是否进入
		    		var footerTop = DOM.footer.getBoundingClientRect().top;
		    		var footerHeight = parseInt(getComputedStyle(DOM.footer)['height']);
		    		var footerInner = browserH - footerTop;

	    			if(footerInner >= 0){
		    			DOM.$summary.css('height', browserH - footerInner + 'px');
		    			DOM.$summary.children('nav').eq(0).css('height',browserH - footerInner - 30 + 'px');
		    		} else {
		    			DOM.$summary.css('height', browserH + 'px');
		    			DOM.$summary.children('nav').eq(0).css('height',browserH - 30 + 'px');
		    		}
	    		};
	    		sumaryHeightFun();
		    	
			}

		};

		uGitpage.init();


	};

	gitbook.events.bind('start', function () {
	 });

	gitbook.events.bind('page.change', function () {
			/**
			 * [uDesign description]
			 * @type {Object}
			 */
	    var uDesign = {
			/**
			 * 初始化
			 * @return {[type]} [description]
			 */
	    	init: function() {
				this.DOMHandler();
				this.leftDirChange();
				this.mobileNav();
				window.onresize = this.mobileNav;

				this.navControl();
				this.buildTag();

				this.addClass();
				this.changeLink();

	    	},

			/**
			 * 获取DOM元素
			 * @type {Object}
			 */
			DOM: {
				"$body": $('body'),
				"$book": $('.book'),	// 文档左侧目录
				"$summary": $('.book-summary'),
				"$bookBody": $('.book-body'),	// 文档右侧主体
				"$bookInner": $('.body-inner')
			},

			DOMHandler: function() {
				var DOM = this.DOM;

				var $container = $('<div class="container"></div>');
				var $containerDiv = $('<div class="container-div"></div>');

				DOM.$book.append( $container );

				$container.append( $containerDiv );
				$containerDiv.append( DOM.$summary );
				$containerDiv.append( DOM.$bookBody );


				// 将超链接放到page-wrapper的最后
				var $pageWrapper = $('.page-wrapper');
				var $prevA = $('.navigation-prev', DOM.$bookBody );
				var $nextA = $('.navigation-next', DOM.$bookBody );

				$pageWrapper.append( $prevA );
				$pageWrapper.append( $nextA );

			    var $html = $('html');
			    $html.css('font-size','62.5%');

			    $('body').css('display','block');
			    document.body.scrollTop = 0;

			    // 设置左侧目录最大高度 : browser height - footerheight - leftmenuMargintop
				var browserH = document.documentElement.clientHeight;
			    DOM.$bookInner.css('min-height',browserH + 'px');

			    // Menu scroll
			    function menuScroll(){
			    	// HEAD Height
		    		var header = document.getElementsByTagName('header')[0];
			    	var banner = document.querySelectorAll('.banner')[0];
			    	var headerHt = parseFloat(getComputedStyle(header)['height']);
			    	var bannerHt = parseFloat(getComputedStyle(banner)['height']);
			    	var headAddHt = headerHt + bannerHt;

			    	var browserH = document.documentElement.clientHeight;

			    	var viewFunc = function() {
			    		var bodyht = document.body.scrollTop || document.documentElement.scrollTop;
				    	var contain = document.querySelectorAll('.book .container')[0];
				    	var leftPadding = parseInt(getComputedStyle(contain)['padding-left']);
				    	var leftBasic = contain.getBoundingClientRect().left;
				    	var leftDis = leftBasic + leftPadding;

				    	var anchorTop = headAddHt - bodyht + 30;

				    	if(bodyht >= headAddHt) {
				    		$('.book-summary').addClass('fix').css('left',leftDis + 'px');
				    		$('#anchors-navbar').css('top', '30px');
				    	} else {
				    		$('.book-summary').removeClass('fix');
				    		$('#anchors-navbar').css('top',anchorTop +'px');
				    	}

			    		// 判断底部进入
				    	var bookInner = document.querySelectorAll('.body-inner')[0];
		    			bookInner.style.minHeight = browserH + 'px';
		    			var bookInnerHeight = parseInt(getComputedStyle(bookInner)['height']);

		    			// 判断footer是否进入
		    			var footer =document.getElementsByTagName('footer')[0];
			    		var footerTop = footer.getBoundingClientRect().top;
			    		var footerHeight = parseInt(getComputedStyle(footer)['height']);
			    		var footerInner = browserH - footerTop;
			    		console.log()
		    			if(footerInner >= 0){
			    			// console.log('scroll in browser');
			    			DOM.$summary.css('height', browserH - footerInner + 'px');
			    			DOM.$summary.children('nav').eq(0).css('height',browserH - footerInner - 30 + 'px');
			    		} else {
			    			// console.log('scroll out browser');
			    			DOM.$summary.css('height', browserH + 'px');
			    			DOM.$summary.children('nav').eq(0).css('height',browserH - 30 + 'px');
			    		}
			    	};

				    document.body.onscroll = viewFunc();
				    window.onresize = viewFunc();
			    }
			    // menuScroll();
			    setInterval(menuScroll, 1);

			    var oH = document.body.offsetHeight;
			    var h = parseInt(oH) - 80;

			    
				},

				/**
				 * 左侧目录修改
				 * @return {[type]} [description]
				 */
				leftDirChange: function() {
					var DOM = this.DOM;

					var $summaryUl = $('ul', DOM.$summary);
					var $dividerLi = $('.divider', DOM.$summaryUl);
					var $dividerLiNext = $('.divider + li', DOM.$summaryUl);

					// 去掉介绍
					$('.summary > li').first().remove();
					
					// 删除下面横线及之后的li
					$dividerLi.remove();
					$dividerLiNext.remove();

					var $summaryAB = $('a b',DOM.$summary);

					// 去掉目录的编号
					$summaryAB.remove();
				},

				/**
				 * gitbook 生成内容的移动端面包导航切换
				 * mobile side menu init
				 * @return {[type]} [description]
				 */
				mobileNav: function() {
					var bodyWidth = document.body.offsetWidth;
					var eleBook = document.querySelectorAll('.book')[0];
					if(bodyWidth<=600) {
						eleBook.classList ? eleBook.classList.remove('with-summary') : eleBook.className.replace(new RegExp('(^|\\b)' + 'with-summary'.split(' ').join('|') + '(\\b|$)', 'gi'), ' ');
						eleBook.classList ? eleBook.classList.remove('with-search') : eleBook.className.replace(new RegExp('(^|\\b)' + 'with-search'.split(' ').join('|') + '(\\b|$)', 'gi'), ' ');
					}
				},

				/**
				 * 右侧导航控制
				 * @return {[type]} [description]
				 */
				navControl: function(){
					var openOffcanvas = $('[data-offcanvas="open"]');
					var closeOffcanvas = $('[data-offcanvas="close"]');
					var offcanvasNav = $('.offcanvas-nav');
					openOffcanvas.click(function(){
						openOffcanvas.addClass('nav-open');
						offcanvasNav.addClass('open');
						$('body').append('<div class="offcanvas-backdrop"></div>');
					});
					closeOffcanvas.click(function(){
						openOffcanvas.removeClass('nav-open');
						offcanvasNav.removeClass('open');
						$('.offcanvas-backdrop').remove();
					});
					$(document).on('click', '.offcanvas-backdrop', function(){
						openOffcanvas.removeClass('nav-open');
						offcanvasNav.removeClass('open');
						$('.offcanvas-backdrop').remove();
					});
				},

		    	/**
		    	 * 创建<style>,<script>标签
		    	 * @return {[type]} [description]
		    	 */
		    	buildTag: function() {
		    		var jsCode = document.querySelectorAll('.jstag');
		    		var cssCode = document.querySelectorAll('.csstag');
		    		var jsTag = document.createElement('script');
		    		var cssTag = document.createElement('style');

		    		for(var i=0, jsLen = jsCode.length; i<jsLen; i++) {
		    			jsTag.innerHTML += jsCode[i].textContent;
		    		}
		    		document.body.appendChild(jsTag);

		    		for(var j=0, cssLen = cssCode.length; j<cssLen; j++ ) {
		    			cssTag.innerHTML += cssCode[j].textContent;
		    		}
		    		document.head.appendChild(cssTag);
		    	},
		    	addClass: function() {
		    		$('.book-summary > nav').addClass('nano');
		    		$('.summary').addClass('nano-content');
		    	},
		    	/**
		    	 * 用于解决gitbook生成的目录没有跳转刷新功能，导致页面加载的插件无法执行
		    	 * @return {[type]} [description]
		    	 */
		    	changeLink: function() {
	                $('.chapter a').on('click', function(){
	                    var aLink = $(this).attr('href').replace(/^\./,'');
	                    var linkPageIndex = window.location.href.lastIndexOf('/');
	                    var lastHref = window.location.href.substr(0,linkPageIndex);
	                    var newHref = window.location.href.substr(0,linkPageIndex) + aLink;
	                    // var newHref = window.location.href.replace(/\/[\w-]+\.html[#]*[\w-]*[\u4e00-\u9fa5]*$/g, aLink); //firefox没识别
	                    window.location.href = newHref;
	                    return false;
	                });
		    	}
		    	
		    }

	    uDesign.init();
	});
});
